File: solutions/fable-1/fable/go/calc.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

type parser struct {
	s   string
	pos int
}

func (p *parser) skip() {
	for p.pos < len(p.s) && (p.s[p.pos] == ' ' || p.s[p.pos] == '\t' || p.s[p.pos] == '\r') {
		p.pos++
	}
}

func (p *parser) peek() byte {
	p.skip()
	if p.pos < len(p.s) {
		return p.s[p.pos]
	}
	return 0
}

func (p *parser) expr() int64 {
	v := p.term()
	for {
		c := p.peek()
		if c == '+' {
			p.pos++
			v += p.term()
		} else if c == '-' {
			p.pos++
			v -= p.term()
		} else {
			return v
		}
	}
}

func (p *parser) term() int64 {
	v := p.factor()
	for {
		c := p.peek()
		if c == '*' {
			p.pos++
			v *= p.factor()
		} else if c == '/' {
			p.pos++
			d := p.factor()
			if d != 0 {
				v /= d // Go integer division truncates toward zero
			}
		} else {
			return v
		}
	}
}

func (p *parser) factor() int64 {
	c := p.peek()
	if c == '(' {
		p.pos++
		v := p.expr()
		if p.peek() == ')' {
			p.pos++
		}
		return v
	}
	var v int64
	for p.pos < len(p.s) && p.s[p.pos] >= '0' && p.s[p.pos] <= '9' {
		v = v*10 + int64(p.s[p.pos]-'0')
		p.pos++
	}
	return v
}

func main() {
	reader := bufio.NewReaderSize(os.Stdin, 1<<20)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()
	for {
		line, err := reader.ReadString('\n')
		if len(line) == 0 && err != nil {
			break
		}
		line = strings.TrimRight(line, "\r\n")
		if strings.TrimSpace(line) == "" {
			if err != nil {
				break
			}
			continue
		}
		p := &parser{s: line}
		fmt.Fprintln(writer, p.expr())
		if err != nil {
			break
		}
	}
}

File: solutions/fable-1/fable/rust/calc.rs
use std::io::{self, BufRead, Write};

struct Parser<'a> {
    s: &'a [u8],
    pos: usize,
}

impl<'a> Parser<'a> {
    fn skip_ws(&mut self) {
        while self.pos < self.s.len() && self.s[self.pos].is_ascii_whitespace() {
            self.pos += 1;
        }
    }

    fn peek(&mut self) -> Option<u8> {
        self.skip_ws();
        self.s.get(self.pos).copied()
    }

    fn expr(&mut self) -> i128 {
        let mut v = self.term();
        loop {
            match self.peek() {
                Some(b'+') => {
                    self.pos += 1;
                    v += self.term();
                }
                Some(b'-') => {
                    self.pos += 1;
                    v -= self.term();
                }
                _ => return v,
            }
        }
    }

    fn term(&mut self) -> i128 {
        let mut v = self.factor();
        loop {
            match self.peek() {
                Some(b'*') => {
                    self.pos += 1;
                    v *= self.factor();
                }
                Some(b'/') => {
                    self.pos += 1;
                    let d = self.factor();
                    if d != 0 {
                        v /= d;
                    }
                }
                _ => return v,
            }
        }
    }

    fn factor(&mut self) -> i128 {
        match self.peek() {
            Some(b'(') => {
                self.pos += 1;
                let v = self.expr();
                if self.peek() == Some(b')') {
                    self.pos += 1;
                }
                v
            }
            Some(b'-') => {
                self.pos += 1;
                -self.factor()
            }
            Some(b'+') => {
                self.pos += 1;
                self.factor()
            }
            Some(c) if c.is_ascii_digit() => {
                let mut v: i128 = 0;
                while self.pos < self.s.len() && self.s[self.pos].is_ascii_digit() {
                    v = v * 10 + (self.s[self.pos] - b'0') as i128;
                    self.pos += 1;
                }
                v
            }
            _ => 0,
        }
    }
}

fn main() {
    let stdin = io::stdin();
    let stdout = io::stdout();
    let mut out = io::BufWriter::new(stdout.lock());
    for line in stdin.lock().lines() {
        let line = match line {
            Ok(l) => l,
            Err(_) => break,
        };
        if line.trim().is_empty() {
            continue;
        }
        let mut p = Parser { s: line.as_bytes(), pos: 0 };
        let v = p.expr();
        writeln!(out, "{}", v).unwrap();
    }
}
